pheget/api/__init__.py

```diff
@@ -2,13 +2,73 @@
 API endpoints (return JSON, not HTML)
 """
 
+import json
+import sqlite3
+import urllib.request
+
 from flask import Blueprint, abort, jsonify, request
 
+from .. import model
 from .format import query_variants
 
 api_blueprint = Blueprint("api", __name__)
 
 
+@api_blueprint.route("/bestvar/<string:symbol>/", methods=["GET"])
+def gene_query(symbol: str):
+    """
+    Given a gene, query an SQL database to find the variant with the strongest association with that gene
+    """
+    # ~UGLY HACK~ If omnisearch did not return a gene_id, then searchbox will fill in "undefined" - immediately abort in that case
+    if symbol == "undefined":
+        abort(404)
+    # Instead of querying locally, I will use omnisearch to get the relevant parameters (chrom, pos, gene_id)
+    url = f"https://portaldev.sph.umich.edu/api/v1/annotation/omnisearch/?q={symbol}&build=GRCh38"
+    omniJson = json.load(urllib.request.urlopen(url))["data"][0]
+    if (
+        "gene_id" in omniJson
+        and "chrom" in omniJson
+        and "start" in omniJson
+        and "end" in omniJson
+    ):
+        conn = sqlite3.connect(model.get_sig_lookup())
+        with conn:
+            try:
+                (
+                    sqlgene_id,
+                    sqlchrom,
+                    sqlpos,
+                    sqlref,
+                    sqlalt,
+                    sqlval,
+                    sqltissue,
+                ) = list(
+                    conn.execute(
+                        f"SELECT * FROM sig WHERE chrom=? AND pos >= ? AND pos <= ? AND gene_id LIKE ? ORDER BY pval LIMIT 1;",
+                        (
+                            "chr" + omniJson["chrom"],
+                            omniJson["start"] - 1000000,
+                            omniJson["end"] + 1000000,
+                            omniJson["gene_id"].split(".")[0] + "._%",
```

Comment:
It looks like the gene ID is a fuzzy match- maybe ENSG versions?

Is there a situation where we would want to store all the versions in the DB? Or should we standardize on a versioned/unversioned identifier?

Reply:
Yes, I found some differences between our current database and omnisearch's ENSG versions. (I should probably have stripped our own (best variant lookup) database's gene_id's of all version numbers.) My database's version numbers comes directly from the data freeze used by GTEx v8.

My database should never have any gene with multiple version numbers - it's just working off a different data freeze than the current omnisearch. Since GTEx v8 is not going to be updated, I consider the version numbers in our database final. (Omnisearch, on the other hand, can be updated in the future, so keeping the fuzzy match will hopefully help with forward compatibility.)

As for going forward, I could re-generate the best variants lookup database by removing all version numbers from the gene_id's, though I think it's useful to know which version the data is based on. And if we are to keep the version numbers, then fuzzy matching seems like the right way to go.

Do you have any particular concerns about any of these approaches (for compatibility/maintainability/etc reasons)?

Reply:
> Do you have any particular concerns about any of these approaches (for compatibility/maintainability/etc reasons)?

Wrt the variant identifier mismatch- keeping track of the small differences is indeed a bit annoying. This is a low grade annoyance for a gtex only project, but if we use this on a different eQTL dataset, it's much more severe because data munging would be spread all through the app. Thus using pheget on a different dataset would require tracking down data manipulation in many many places.

In that world, I could imagine normalizing identifiers to be something that happened during an initial intake step, or, at very least, a helper function that makes it clear what is going. We've talked about separating "data" from "views" in other contexts; the "models" module could certainly be beefed up for that purpose.